### PR TITLE
feat(matchedData): add generic types for return

### PR DESCRIPTION
**Pull request by otorrillas**
## Description

As brought up in https://github.com/express-validator/express-validator/issues/1041, it's currently _hard_ to type the result of `parsedData`. It comes with all fields set as `any` unless we do a [type assertion](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions) via the `as` keyword.
```ts
const { something } = parsedData(req);
// something has any type

const { something } = parsedData(req) as { something: string };
// something has string type
```

However, this syntax _overrides_ all type-checking. To mitigate that, this PR adds an optional [generic type](https://www.typescriptlang.org/docs/handbook/2/generics.html) to the `parsedData` function signature, so it becomes:

```ts
function matchedData<T = Record<string, any>>(
  req: Request,
  options: Partial<MatchedDataOptions> = {},
): T {
	// ...
}
```

Allowing users to keep using `matchedData` with the same signature as before:
```ts
const { something } = parsedData(req);
// something has any type
```

but to provide a type signature of the expected return (if it's been validated) if they'd like:
```ts
const { something } = parsedData<{ something: string }>(req);
// something has string type
```

There are **no breaking changes**, since `matchedData` still keeps `Record<string, any>` as the default return type.

> [!NOTE]
> While types might not be strictly accurate depending on how the generic type is created, I believe it's better than the current solution (`Record<string, any>`).
> As suggested in https://github.com/express-validator/express-validator/issues/1041#issuecomment-1150448639, though, I believe a more type could be provided from the library depending on the options provided (e.g. with `includeOptionals: true`) via [type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types). I'm open to exploring that direction further if you (maintainers) believe it might be worth implementing 👍 

## To-do list

- [X] I have added tests for what I changed.

- [X] This pull request is ready to merge.

### CODE REVIEW

**Review by arafatkn — APPROVED** [review]
Looking good.

File: docs/api/misc.md [hunk 1]
```diff
@@ -76,3 +76,35 @@ type Location = 'body' | 'cookies' | 'headers' | 'params' | 'query';
 
 Represents one of the request locations: `req.body`, `req.cookies`, `req.headers`, `req.params` and
 `req.query`.
+
+### `matchedData`
```
Comment (gustavohenke) [inline docs/api/misc.md:80]: Can you move these next to `matchedData`'s own documentation, in https://github.com/express-validator/express-validator/blob/master/docs/api/matched-data.md? Maybe under a "TypeScript usage" header.

Thanks!

**Review by gustavohenke — CHANGES_REQUESTED** [review]

File: docs/api/misc.md (hunk 1 above)
Comment (otorrillas) [reply docs/api/misc.md:80]: Sure thing! Thanks!

**Review by gustavohenke — APPROVED** [review]
thanks, this is an excellent enhancement 😃

File: src/matched-data.ts [hunk 2]
```diff
@@ -49,8 +49,7 @@ export function matchedData(
     .filter(validityFilter)
     .map(field => field.instance)
     .filter(locationFilter)
-    .reduce((state, instance) => _.set(state, instance.path, instance.value), {})
-    .valueOf();
```
Comment (gustavohenke) [inline src/matched-data.ts:53]: Had to make a fix here; missed how [`_(...).reduce()` automatically unwraps the result value](https://lodash.com/docs/4.17.15#lodash), meaning that this `valueOf()` was incorrect here.

**Pull request merged**